src/main/scala/mesosphere/marathon/core/health/impl/AppHealthCheckActor.scala

```diff
@@ -0,0 +1,236 @@
+import scala.collection.mutable
+
+/**
+  * Actor computing instance healthiness of all instances of each application
+  * in order to send `InstanceHealthChanged` events to the event bus when the healthiness
+  * of a given instance changes. It basically aggregates all statuses of a given instance.
+  *
+  * There are 3 possible healthiness states:
+  * - `Unknown` if at least one health check status of the instance is unknown
+  * - `Unhealthy` if all statuses of the instance are known but at least one is failing
+  * - `Healthy` if all statuses of the instance are known and all are healthy
+  */
+object AppHealthCheckActor {
+  case class ApplicationKey(appId: PathId, version: Timestamp)
+  case class InstanceKey(applicationKey: ApplicationKey, instanceId: Instance.Id)
+
+  def props(eventBus: EventStream): Props = Props(new AppHealthCheckActor(eventBus))
+
+  /**
+    * Actor command adding an health check definition for a given application so that the actor knows which health check
+    * statuses are know and which are not for a given instance.
+    * @param appKey The application key of the application for which the health check is defined
+    * @param healthCheck The definition of the health check to add
+    */
+  case class AddHealthCheck(appKey: ApplicationKey, healthCheck: HealthCheck)
+
+  /**
+    * Actor command removing a health check definition for a given application.
+    * This method also purges the statuses of health checks of all instances that ran it.
+    * @param appKey The application key of the application for which the health check is removed
+    * @param healthCheck The definition of the health check to remove
+    */
+  case class RemoveHealthCheck(appKey: ApplicationKey, healthCheck: HealthCheck)
+
+  /**
+    * Actor command purging several health check statuses.
+    * @param toPurge The list of health check statuses to purge.
+    */
+  case class PurgeHealthCheckStatuses(toPurge: Seq[(InstanceKey, HealthCheck)])
+
+  /**
+    * Actor command updating the status of an health check of a given instance of an application. If the instance status
+    * has changed, it sends a `InstanceHealthChanged` event to the event bus.
+    * @param appKey The application key for which the health check status is updated
+    * @param healthCheck The health check for which the status is updated
+    * @param healthState The new state of the health check
+    */
+  case class HealthCheckStatusChanged(
+      appKey: ApplicationKey,
+      healthCheck: HealthCheck, healthState: Health)
+
+  /**
+    * This proxy class is the implementation of the actor. It has been created to be used in performance benchmarks.
+    * Beware this class is NOT thread safe.
+    */
+  class AppHealthCheckProxy extends StrictLogging {
+    /**
+      * Map of health check definitions of all applications
+      */
+    private[impl] val healthChecks: mutable.Map[ApplicationKey, Set[HealthCheck]] = mutable.Map.empty
```

Comment:
Do we care about "thread safety" here? I think we should. The docs say:

> if an actor makes changes to its internal state while processing a message, and accesses that state while processing another message moments later. It is important to realize that with the actor model you don't get any guarantee that the same thread will be executing the same actor for different messages.

Reply:
On the other hand, the docs say:

> In layman's terms this means that changes to internal fields of the actor are visible when the next message is processed by that actor. So fields in your actor need not be volatile or equivalent.

Reply:
It seems it is addressed by the "happens-before" rules. In this case, please ignore me.

Reply:
There is a comment at the top of this class, "BEWARE this class is NOT thread safe". This, plus the fact that this class is used by an actor, resolves all concerns I have.